pkg/streamingpromql/operators/aggregations/limitklimitratio/query.go

```diff
@@ -0,0 +1,431 @@
+}
+
+func (q *limitKQueryInternalImpls[p]) enableGroupStepCounter() bool {
+	return true
+}
+
+func (q *limitKQueryInternalImpls[p]) newLimitParam(ctx context.Context, annotations *annotations.Annotations, memoryConsumptionTracker *limiter.MemoryConsumptionTracker, stepCount int, param types.ScalarOperator) (param, error) {
+	var err error
+	q.param, err = newKParam(ctx, memoryConsumptionTracker, stepCount, param)
+	if err != nil {
+		return nil, err
+	}
+	return q.param, nil
+}
+
+func (q *limitKQueryInternalImpls[p]) maxPossibleSeries(g *queryGroup) int {
+	// consider the largest k vs the number of series observed for this group
+	return int(min(q.param.kMax, int64(g.seriesCount)))
+}
+
+type limitRatioQueryInternalImpls[p ratioParam] struct {
+	param *ratioParam
+}
+
+func (q *limitRatioQueryInternalImpls[p]) discardSeries(metadata types.SeriesMetadata, data types.InstantVectorSeriesData, g *queryGroup) bool {
+	return q.param.rStepInvariant && !ratiosampler.AddRatioSample(q.param.r[0], &promql.Sample{Metric: metadata.Labels})
+}
+
+func (q *limitRatioQueryInternalImpls[p]) discardPointAtStep(metadata types.SeriesMetadata, step int, g *queryGroup) bool {
+	r := q.param.r[step]
+	return !ratiosampler.AddRatioSample(r, &promql.Sample{Metric: metadata.Labels})
+}
+
+func (q *limitRatioQueryInternalImpls[p]) enableGroupStepCounter() bool {
+	return false
+}
+
+func (q *limitRatioQueryInternalImpls[p]) isAllFilled(step int, g *queryGroup) bool {
+	return false
+}
+
+func (q *limitRatioQueryInternalImpls[p]) newLimitParam(ctx context.Context, annotations *annotations.Annotations, memoryConsumptionTracker *limiter.MemoryConsumptionTracker, stepCount int, param types.ScalarOperator) (param, error) {
+	var err error
+	q.param, err = newRatioParam(ctx, annotations, memoryConsumptionTracker, stepCount, param)
+	if err != nil {
+		return nil, err
+	}
+	return q.param, nil
+}
+
+func (q *limitRatioQueryInternalImpls[p]) maxPossibleSeries(g *queryGroup) int {
+	return g.seriesCount
+}
+
+type Query[p param] struct {
+	impls queryInternalImpls[p]
+	param param
+
+	Param                    types.ScalarOperator
+	Inner                    types.InstantVectorOperator
+	TimeRange                types.QueryTimeRange
+	Grouping                 []string // If this is a 'without' aggregation, New will ensure that this slice contains __name__.
+	Without                  bool
+	MemoryConsumptionTracker *limiter.MemoryConsumptionTracker
+
+	expressionPosition posrange.PositionRange
+	annotations        *annotations.Annotations
+
+	// index by output series id - populated via the SeriesMetadata() and then returned in the NextSeries()
+	values []*types.InstantVectorSeriesData
+
+	// the next output series to be returned in NextSeries()
+	nextSeriesIndex int
+}
+
+func (t *Query[p]) SeriesMetadata(ctx context.Context, matchers types.Matchers) ([]types.SeriesMetadata, error) {
+	var err error
+	functionFailedToComplete := true
```

Comment:
I wouldn't bother with this - it adds complexity and it's fine not to return everything to the pool if an error occurs.